core/src/com/put/urbantraffic/Car.java

```diff
@@ -1,55 +1,137 @@
-    private final List<Node> path;
+    private final Lane startLane;
+    private final Lane endLane;
+    private Node startNode;
+    private Node endNode;
+
+    private List<Node> path;
+
     private Node currentNode;
     private Node nextNode;
-    private int xPos;
-    private int yPos;
-    private int nodePercentage = 0;
+    private Node actualNode;
+    private Lane currentLane;
+    private float nodePercentage = 0.0f;
+
     private RideStatus status = RideStatus.WAITING;
 
-    public Car(Node startNode, Node endNode, List<Node> path) {
-        this.startNode = startNode;
-        this.endNode = endNode;
-        this.path = path;
-        this.currentNode = path.get(0);
+    private List<Crossing> crossingList;
+    private List<Lane> lanesList;
+    CityGraph.PathWithTime calculatedPath;
+
+    private Way way;
+
+    public enum Way {
+        TOP, RIGHT, BOTTOM, LEFT
+    }
+
+    public Car(Lane startLane, Lane endLane) {
+        this.startLane = startLane;
+        this.endLane = endLane;
+
+        this.startNode = this.startLane.getMiddlePoint();
+        this.endNode = this.endLane.getMiddlePoint();
+
+        this.path = generatePath(startLane, endLane);
+
+        this.currentNode = new Node(path.get(0).getX(), path.get(0).getY());
+        this.currentLane = lanesList.get(0);
+        this.actualNode = new Node(path.get(0).getX(), path.get(0).getY());
         this.nextNode = path.get(1);
-        this.xPos = startNode.getX();
-        this.yPos = startNode.getY();
+
+        this.way = calculateWay(this.currentNode, this.nextNode);
+
+        System.out.println(this.path);
+    }
+
+    private List<Node> generatePath(Lane startLane, Lane endLane) {
+        calculatedPath = UrbanTrafficFlowSimulation.paths[startLane.getId()][endLane.getId()];
+        this.crossingList = calculatedPath.getCrossings();
+
+        this.lanesList = new ArrayList<>(Collections.singletonList(this.startLane));
+        this.lanesList.addAll(calculatedPath.getLanes());
+        this.lanesList.add(endLane);
+
+        System.out.println(calculatedPath);
+
+        List<Node> path = new ArrayList<>(Collections.singletonList(this.startNode));
+        if (this.crossingList.size() == 1)
+            path.add(new Node(this.crossingList.get(0).getX(), this.crossingList.get(0).getY()));
+        else
+            for (Lane lane : calculatedPath.getLanes()) {
+                for (Node node : lane.getNodeList()) {
+                    if (!node.equals(path.get(path.size() - 1))) {
+                        path.add(node);
+                    }
+                }
+            }
+        path.add(this.endNode);
+
+        System.out.println("NODES: " + path.size() + " ROADS: " + this.lanesList.size());
+        return path;
     }
 
     public void moveCar() {
         if (status != RideStatus.FINISH) {
-            nodePercentage += 1;
+            ArrayList<Node> nodeList = new ArrayList<>(currentLane.getNodeList());
```

Comment:
why are we copying the list here? We don't modify it, so nothing wrong can happen, right?

Reply:
Yes, you are right, I will fix it 👷‍♂️